url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/TMAProtocol.java

```diff
@@ -0,0 +1,351 @@
+		EqualChallengesGivenDifferentGiven[i] = tmp1.multiply(tmp8.multiply(tmp5));
+		return EqualChallengesGivenDifferentGiven[i];
+	}
+
+	BigDecimal[] distanceFraud = new BigDecimal[257];
+
+	@Override
+	public BigDecimal getDistanceFraudProbability() {
+		//return distanceFraud(n, 0);
+		if (n == 0) return BigDecimal.ONE;
+		return new BigDecimal(distanceFraudFromDisk2(n));
+	}
+	
+	
+	public BigDecimal distanceFraud(int n, int q) {
+		if (n == 0) return new BigDecimal(1);
+		if (distanceFraud[n] != null) return distanceFraud[n];
+		BigDecimal tmp1 = distanceFraud(n-1, q);
+		BigDecimal tmp2 = computeFGivenD(n-1);
+		tmp2 = tmp2.multiply(new BigDecimal(0.5));
+		tmp1 = tmp1.multiply(new BigDecimal(0.25));
+		distanceFraud[n] = tmp2.add(tmp1);
+		return distanceFraud[n];
+	}
+
+
+	BigDecimal[] FGivenD = new BigDecimal[257];
+
+	
+	private BigDecimal computeFGivenD(int i) {
+		if (i == 0) return BigDecimal.ONE;
+		if (FGivenD[i] != null) return FGivenD[i];
+		BigDecimal tmp2 = distanceFraud(i-1, 0);
+		tmp2 = tmp2.multiply(new BigDecimal("0.125"));
+		BigDecimal tmp5 = computeFGivenD(i-1);
+		//tmp5 = tmp5.multiply(new BigDecimal("0.53125"));
+		tmp5 = tmp5.multiply(new BigDecimal("0.5"));
+		tmp5 = tmp5.add(tmp2);
+		FGivenD[i] = tmp5;
+		return FGivenD[i];
+	}
+
+
+	@Override
+	public BigDecimal getTerroristFraudProbability() {
+		return ONE;
+	}
+
+	@Override
+	public boolean hasFinalSlowPhase() {
+		return false;
+	}
+
+	@Override
+	public int getSizeOfTheChannel() {
+		return 1;
+	}
+
+	@Override
+	public boolean lackSecurityProof() {
+		return true;
+	}
+
+	@Override
+	public int getYearOfPublication() {
+		return 2014;
+	}
+
+	@Override
+	public DBProtocol[] getInstances() {
+		DBProtocol[] result = new DBProtocol[MAX_N];
+		for (int i = 0; i < MAX_N; i++){
+			result[i] = new TMAProtocol();
+			result[i].setNumberOfRounds(i+1);
+		}
+		return result;
+	}
+
+	@Override
+	public String getIdentifier() {
+		return "TMA-{"+n+"}";
+	}
+
+	@Override
+	public int getCryptoCalls() {
+		return 1;
+	}
+
+	/** Apr 20, 2010 TnT Comment
+	 * El valor del distance fraud esta en disco. Depende del valor de a, el archivo se llama "new_tree_a.dat". 
+	 * Y entonces en este archivo podras encontrar la ronda a la que te encuentras.*/
+	private double distanceFraudFromDisk2(int n){
+		if (distances2 == null) {
+			try {
+				fillDistance2(257);
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+				throw new RuntimeException(e);
+			}
+		}
+		return distances2[n];
+	}
+	private double[] distances2;
+	/*La informacion sobre el distance esta en un fichero de nombre "d_FDA-"+n*/
+	private void fillDistance2(int n) throws IOException {
+		//System.out.println("loaded file for distance fraud of the New protocol");
+		String tmp;
+		String[] split;
+		int r;
+		distances2 = new double[n];
+		File f = new File(getDistanceFileNameOfValues());
+		BufferedReader reader = new BufferedReader(new FileReader(f));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 247, 248, 249. The resource is closed at line 254. There are other execution paths that do not contain closure statements, for example, when BufferedReader.ready throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).